specs/multi-url-support.md

```diff
@@ -0,0 +1,112 @@
+# Support for any URLs via the Microsoft Graph Developer Proxy
+
+Some applications are using an hybrid approach when it comes to consuming APIs. Some APIs are available on Microsoft Graph while some are not. To enable developers to fully test their scenarios with the Microsoft Graph Developer Proxy, the proxy should support acting as a reverse proxy for more than just one URL at a time, and include URLs that are outside of the Microsoft Graph. This can also be used by internal teams to test their APIs on internal endpoints.
+
+## History
+
+| Version | Date | Comments | Author |
+| ------- | -------- | ----- | --- |
+| 1.0 | 2022-12-09 | Initial specifications | @sebastienlevert
+
+
+## Configuration file
+
+A configuration file should be available for developers who want to support scenarios outside of Microsoft Graph. The file should be at the root of the proxy working folder and named `config.json`. This file should have default RegEx values for all supported Microsoft Graph clouds and SharePoint URLs.
+
+This configuration can also be extended by "any" URLs to support 1P endpoints, vanity URLs, custom services, etc.
+
+| Property | Type | Comments |
+| ------- | -------- | ----- |
+| id | String (unique) | This can be any string value but needs to be unique |
+| regEx | String | String value of the RegEx to use |
+| isWildcard | Boolean | If the value is `true`, the RegEx gets an appended wildcard token. If `false`, only the RegEx value should be considered |
+
+> **Note**
+> These RegEx need to be validated in the context of the App
+
+```json
+{

```

Comment:
I suggest to make it easier and use the following structure to define the list of hosts to watch:

```json
{
  "hostsToWatch": [
    "graph.microsoft.com",
    "graph.microsoft.us",
    ...
    "*.sharepoint.com"
  ]
}
```

Rather than using regex, we'd use the existing wildcard-based pattern matching, that we have already implemented in matching URLs for mock responses, allowing us to reuse existing code, and implement this quicker. With the existing pattern matching, we check for the presence of `*` in a string. If we find it, we use regex matching replacing `*` with `.*` (basically any character). If we don't, we use a string equality comparison to determine the match.

Requiring users to build regex makes it harder to use the proxy. Writing regex is hard. Especially, because there's no easy way for folks to debug the queries, it will force them to use other tools, further adding complexity.
Writing good regex is hard, and this is important because we'll be using these patterns to match against every request we intercept and then again to check if the response should be mocked. So if the regex is written badly, it might significantly slow down the proxy.

Using a simple list of hosts instead of regex and complex objects has another benefit. When we add support for importing mocks from a HAR file, we could automatically detect host names that we don't have in the list and add them to ensure that these requests are properly captured as well. If we work with regex and objects , there's a good chance we'll end up with duplicate definitions which could lead to undesired side-effects and hard to debug issues.

Reply:
I like this reasoning to keep it simple. You're absolutely right about RegEx, it's easy to get it wrong because it's arcane and varies depending on the run time. I really like the idea of supporting simple wildcard matching.